File: modules/resin/src/com/caucho/server/resin/ResinArgs.java
/*
 * Copyright (c) 1998-2012 Caucho Technology -- all rights reserved
 *
 * This file is part of Resin(R) Open Source
 *
 * Each copy or derived work must preserve the copyright notice and this
 * notice unmodified.
 *
 * Resin Open Source is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Resin Open Source is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE, or any warranty
 * of NON-INFRINGEMENT.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Resin Open Source; if not, write to the
 *
 *   Free Software Foundation, Inc.
 *   59 Temple Place, Suite 330
 *   Boston, MA 02111-1307  USA
 *
 * @author Scott Ferguson
 */

package com.caucho.server.resin;

import java.net.Socket;
import java.net.URL;
import java.util.ArrayList;
import java.util.Date;
import java.util.logging.Level;
import java.util.logging.Logger;

import com.caucho.VersionFactory;
import com.caucho.config.ConfigException;
import com.caucho.log.EnvironmentStream;
import com.caucho.log.RotateStream;
import com.caucho.server.util.JniCauchoSystem;
import com.caucho.util.L10N;
import com.caucho.vfs.Path;
import com.caucho.vfs.QJniServerSocket;
import com.caucho.vfs.QServerSocket;
import com.caucho.vfs.Vfs;
import com.caucho.vfs.WriteStream;

/**
 * The parsed Resin command-line arguments
 */
public class ResinArgs
{
  private static final Logger log = Logger.getLogger(ResinArgs.class.getName());
  private static final L10N L = new L10N(ResinArgs.class);

  private String _serverId;

  private Path _resinHome;
  private Path _rootDirectory;
  private Path _dataDirectory;
  private Path _licenseDirectory;
  
  private Path _userProperties;

  private String _resinConf;
  
  private Socket _pingSocket;
  
  private boolean _isElasticServer;
  private String _homeCluster;
  private String _elasticServerAddress;
  private int _elasticServerPort;
  
  private ArrayList<BoundPort> _boundPortList
    = new ArrayList<BoundPort>();

  private boolean _isOpenSource;
  
  private String _stage = "production";
  private String _mode = "default";
  
  private boolean _isDumpHeapOnExit;
  
  public ResinArgs()
  {
    this(new String[0]);
  }
  
  public ResinArgs(String []args)
    throws ConfigException
  {
    try {
      initEnvironmentDefaults();
    
      parseCommandLine(args);
    } catch (Exception e) {
      throw ConfigException.create(e);
    }
  }

  private void initEnvironmentDefaults()
  {
    String resinHome = System.getProperty("resin.home");

    if (resinHome != null)
      _resinHome = Vfs.lookup(resinHome);
    else
      _resinHome = Vfs.getPwd();

    _rootDirectory = _resinHome;

    // server.root backwards compat
    String resinRoot = System.getProperty("server.root");

    if (resinRoot != null)
      _rootDirectory = Vfs.lookup(resinRoot);

    // resin.root backwards compat
    resinRoot = System.getProperty("resin.root");

    if (resinRoot != null)
      _rootDirectory = Vfs.lookup(resinRoot);

    try {
      URL.setURLStreamHandlerFactory(ResinURLStreamHandlerFactory.create());
    } catch (java.lang.Error e) {
      //operation permitted once per jvm; catching for harness.
    }
    
    _userProperties = Vfs.lookup(System.getProperty("user.home") + "/.resin");
  }
  
  public void setServerId(String serverId)
  {
    if ("".equals(serverId)) {
      serverId = "default";
    }
    
    _serverId = serverId;
  }

  public String getServerId()
  {
    return _serverId;
  }

  public Path getResinHome()
  {
    return _resinHome;
  }
  
  /**
   * Gets the root directory.
   */
  public Path getRootDirectory()
  {
    return _rootDirectory;
  }
  
  public void setRootDirectory(Path root)
  {
    _rootDirectory = root;
  }

  public Path getLicenseDirectory()
  {
    return _licenseDirectory;
  }

  public void setLicenseDirectory(Path licenseDirectory)
  {
    _licenseDirectory = licenseDirectory;
  }

  /**
   * Gets the root directory.
   */
  public Path getDataDirectory()
  {
    return _dataDirectory;
  }
  
  /**
   * Sets the root directory.
   */
  public void setDataDirectory(Path path)
  {
    _dataDirectory = path;
  }
  
  public Socket getPingSocket()
  {
    return _pingSocket;
  }
  
  public void setOpenSource(boolean isOpenSource)
  {
    _isOpenSource = isOpenSource;
  }
  
  public boolean isOpenSource()
  {
    return _isOpenSource;
  }

  /**
   * The configuration file used to start the server.
   */
  public String getResinConf()
  {
    return _resinConf;
  }
  
  public void setResinConf(String resinConf)
  {
    _resinConf = resinConf;
  }
  
  public Path getUserProperties()
  {
    return _userProperties;
  }
  
  public String getMode()
  {
    return _mode;
  }

  public Path getResinConfPath()
  {
    Path pwd = Vfs.lookup();

    Path resinConf = null;
    
    String resinConfFile = getResinConf();
    
    if (resinConfFile != null) {
      if (log.isLoggable(Level.FINER))
        log.finer(this + " looking for conf in " +  pwd.lookup(resinConfFile));

      resinConf = pwd.lookup(resinConfFile);
    }
    else if (pwd.lookup("conf/resin.xml").canRead())
      resinConfFile = "conf/resin.xml";
    else { // backward compat
      resinConfFile = "conf/resin.conf";
    }
    
    Path rootDirectory = getRootDirectory();

    if (resinConf == null || ! resinConf.exists()) {
      if (log.isLoggable(Level.FINER))
        log.finer(this + " looking for conf in "
                  +  rootDirectory.lookup(resinConfFile));

      resinConf = _rootDirectory.lookup(resinConfFile);
    }

    if (! resinConf.exists() && ! _resinHome.equals(_rootDirectory)) {
      if (log.isLoggable(Level.FINER))
        log.finer(this + " looking for conf in "
                  +  _resinHome.lookup(resinConfFile));

      resinConf = _resinHome.lookup(resinConfFile);
    }

    // for error messages, show path relative to rootDirectory
    if (! resinConf.exists())
      resinConf = rootDirectory.lookup(resinConfFile);

    return resinConf;
  }
  
  /**
   * Returns the bound port list.
   */
  public ArrayList<BoundPort> getBoundPortList()
  {
    return _boundPortList;
  }
  
  /**
   * Returns the stage to start with.
   */
  public String getStage()
  {
    return _stage;
  }
  
  public void setStage(String stage)
  {
    _stage = stage;
  }
  
  public void setElastic(boolean isElastic)
  {
    _isElasticServer = isElastic;
  }
  
  public boolean isElasticServer()
  {
    return _isElasticServer;
  }
  
  public int getElasticServerPort()
  {
    return _elasticServerPort;
  }
  
  public String getElasticServerAddress()
  {
    return _elasticServerAddress;
  }
  
  public void setHomeCluster(String homeCluster)
  {
    _homeCluster = homeCluster;
  }
  
  public String getHomeCluster()
  {
    return _homeCluster;
  }
  
  public String getServerAddress()
  {
    return _elasticServerAddress;
  }
  
  public void setServerAddress(String address)
  {
    _elasticServerAddress = address;
  }
  
  public int getServerPort()
  {
    return _elasticServerPort;
  }
  
  public void setServerPort(int port)
  {
    _elasticServerPort = port;
  }
  
  public String getUser()
  {
    return null;
  }
  
  public String getPassword()
  {
    return null;
  }
  
  public boolean isDumpHeapOnExit()
  {
    return _isDumpHeapOnExit;
  }

  private void parseCommandLine(String []argv)
    throws Exception
  {
    int len = argv.length;
    int i = 0;

    while (i < len) {
      // RandomUtil.addRandom(argv[i]);
      
      String arg = argv[i];
      
      if (arg.startsWith("-") && ! arg.startsWith("--"))
        arg = "-" + arg;

      if (i + 1 < len && arg.equals("--stdout")) {
        Path path = Vfs.lookup(argv[i + 1]);

        RotateStream stream = RotateStream.create(path);
        stream.init();
        WriteStream out = stream.getStream();
        out.setDisableClose(true);

        EnvironmentStream.setStdout(out);

        i += 2;
      }
      else if (i + 1 < len && arg.equals("--stderr")) {
        Path path = Vfs.lookup(argv[i + 1]);

        RotateStream stream = RotateStream.create(path);
        stream.init();
        WriteStream out = stream.getStream();
        out.setDisableClose(true);

        EnvironmentStream.setStderr(out);

        i += 2;
      }
      else if (i + 1 < len && arg.equals("--conf")) {
        _resinConf = argv[i + 1];
        i += 2;
      }
      else if (arg.equals("--log-directory")) {
        i += 2;
      }
      else if (arg.equals("--config-server")) {
        i += 2;
      }
      else if (arg.equals("--dump-heap-on-exit")) {
        _isDumpHeapOnExit = true;

        i += 1;
      }
      else if (i + 1 < len && arg.equals("--server")) {
        _serverId = argv[i + 1];
   
        if (_serverId.equals(""))
          _serverId = "default";
   
        i += 2;
      }
      else if (arg.equals("--resin-home")) {
        _resinHome = Vfs.lookup(argv[i + 1]);

        i += 2;
      }
      else if (arg.equals("--root-directory")
               || arg.equals("--resin-root")
               || arg.equals("--server-root")) {
        _rootDirectory = _resinHome.lookup(argv[i + 1]);

        i += 2;
      }
      else if (arg.equals("--data-directory")) {
        _dataDirectory = Vfs.lookup(argv[i + 1]);

        i += 2;
      }
      else if (arg.equals("--elastic-server")) {
        _isElasticServer = true;
        i += 1;
      }
      else if (i + 1 < len
               && (arg.equals("--elastic-server-address")
                   || arg.equals("--server-address"))) {
        _elasticServerAddress = argv[i + 1];
   
        i += 2;
      }
      else if (i + 1 < len
               && (arg.equals("--elastic-server-port")
                   || arg.equals("--elastic-server-port"))) {
        _elasticServerPort = Integer.parseInt(argv[i + 1]);
   
        i += 2;
      }
      else if (arg.equals("--elastic-dns")) {
        i += 1;
      }
      else if (arg.equals("--service")) {
        JniCauchoSystem.create().initJniBackground();
        // windows service
        i += 1;
      }
      else if (i + 1 < len
               && (arg.equals("--cluster")
                   || arg.equals("--join-cluster"))) {
        _homeCluster = argv[i + 1];
   
        i += 2;
      }
      else if (arg.equals("--version")) {
        System.out.println(VersionFactory.getFullVersion());
        System.exit(66);
      }
      else if (arg.equals("--watchdog-port")) {
        // watchdog
        i += 2;
      }
      else if (arg.equals("--socketwait")
               || arg.equals("--pingwait")) {
        int socketport = Integer.parseInt(argv[i + 1]);

        Socket socket = null;
        for (int k = 0; k < 15 && socket == null; k++) {
          try {
            socket = new Socket("127.0.0.1", socketport);
          } catch (Throwable e) {
            System.out.println(new Date());
            e.printStackTrace();
          }

          if (socket == null)
            Thread.sleep(1000);
        }

        if (socket == null) {
          System.err.println("Can't connect to parent process through socket " + socketport);
          System.err.println("Resin needs to connect to its parent.");
          System.exit(1);
        }

        /*
        if (argv[i].equals("-socketwait") || argv[i].equals("--socketwait"))
          _waitIn = socket.getInputStream();
        */

        _pingSocket = socket;

        //socket.setSoTimeout(60000);

        i += 2;
      }
      else if ("-port".equals(argv[i]) || "--port".equals(argv[i])) {
        int fd = Integer.parseInt(argv[i + 1]);
        String addr = argv[i + 2];
        if ("null".equals(addr))
          addr = null;
        int port = Integer.parseInt(argv[i + 3]);

        _boundPortList.add(new BoundPort(QJniServerSocket.openJNI(fd, port),
                                         addr,
                                         port));

        i += 4;
      }
      else if ("start".equals(argv[i])
               || "start-all".equals(argv[i])
               || "start-cloud".equals(argv[i])
               || "restart".equals(argv[i])) {
        JniCauchoSystem.create().initJniBackground();
        i++;
      }
      else if (argv[i].equals("-verbose")
               || argv[i].equals("--verbose")) {
        i += 1;
      }
      else if (argv[i].equals("gui")) {
        i += 1;
      }
      else if (argv[i].equals("console")) {
        i += 1;
      }
      else if (argv[i].equals("watchdog")) {
        i += 1;
      }
      else if (argv[i].equals("start-with-foreground")) {
        i += 1;
      }
      else if (argv[i].equals("-fine")
               || argv[i].equals("--fine")) {
        i += 1;
      }
      else if (argv[i].equals("-finer")
               || argv[i].equals("--finer")) {
        i += 1;
      }
      else if (argv[i].startsWith("-D")
               || argv[i].startsWith("-J")
               || argv[i].startsWith("-X")) {
        i += 1;
      }
      else if ("-stage".equals(argv[i])
               || "--stage".equals(argv[i])) {
        _stage = argv[i + 1];
        i += 2;
      }
      else if ("-preview".equals(argv[i])
               || "--preview".equals(argv[i])) {
        _stage = "preview";
        i += 1;
      }
      else if ("-debug-port".equals(argv[i])
               || "--debug-port".equals(argv[i])) {
        i += 2;
      }
      else if ("-jmx-port".equals(argv[i])
               || "--jmx-port".equals(argv[i])) {
        i += 2;
      }
      else if ("-jmx-port".equals(argv[i])
               || "--jmx-port".equals(argv[i])) {
        i += 2;
      }
      else if ("--user-properties".equals(arg)) {
        _userProperties = Vfs.lookup(argv[i + 1]);
        
        i += 2;
      }
      else if ("--mode".equals(arg)) {
        _mode = argv[i + 1];
        
        i += 2;
      }
      else {
        System.out.println(L.l("unknown argument '{0}'", arg));
        System.out.println();
        usage();
        System.exit(66);
      }
    }
  }

  private static void usage()
  {
    System.err.println(L.l("usage: bin/resin.sh [-options] [start | stop | restart]"));
    System.err.println(L.l(""));
    System.err.println(L.l("where options include:"));
    System.err.println(L.l("   -conf <file>          : select a configuration file"));
    System.err.println(L.l("   -data-directory <dir> : select a resin-data directory"));
    System.err.println(L.l("   -log-directory <dir>  : select a logging directory"));
    System.err.println(L.l("   -resin-home <dir>     : select a resin home directory"));
    System.err.println(L.l("   -root-directory <dir> : select a root directory"));
    System.err.println(L.l("   -server <id>          : select a <server> to run"));
    System.err.println(L.l("   -watchdog-port <port> : override the watchdog-port"));
    System.err.println(L.l("   -verbose              : print verbose starting information"));
    System.err.println(L.l("   -preview              : run as a preview server"));
  }

  static class DynamicServer {
    private final String _cluster;
    private final String _address;
    private final int _port;

    DynamicServer(String cluster, String address, int port)
    {
      _cluster = cluster;
      _address = address;
      _port = port;
    }

    String getCluster()
    {
      return _cluster;
    }

    String getAddress()
    {
      return _address;
    }

    int getPort()
    {
      return _port;
    }
  }
  
  static class BoundPort {
    private QServerSocket _ss;
    private String _address;
    private int _port;

    BoundPort(QServerSocket ss, String address, int port)
    {
      if (ss == null)
        throw new NullPointerException();

      _ss = ss;
      _address = address;
      _port = port;
    }

    public QServerSocket getServerSocket()
    {
      return _ss;
    }

    public int getPort()
    {
      return _port;
    }

    public String getAddress()
    {
      return _address;
    }
  }
}

File: modules/csharp/src/common/ResinArgs.cs
/*
 * Copyright (c) 1998-2015 Caucho Technology -- all rights reserved
 *
 * This file is part of Resin(R) Open Source
 *
 * Each copy or derived work must preserve the copyright notice and this
 * notice unmodified.
 *
 * Resin Open Source is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * Resin Open Source is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE, or any warranty
 * of NON-INFRINGEMENT.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Resin Open Source; if not, write to the
 *
 *   Free Software Foundation, Inc.
 *   59 Temple Place, Suite 330
 *   Boston, MA 02111-1307  USA
 *
 * @author Alex Rojkov
 */
using System;
using System.Collections.Specialized;
using System.Text;

namespace Caucho
{
  public class ResinArgs
  {
    public bool IsVerbose { get; private set; }
    public bool IsService { get; private set; }

    //used with -install / -remove.
    //value of true indicates that the process is started
    //from resin.exe with elevated privileges
    public String JavaExe { get; private set; }
    public String JavaHome { get; private set; }
    public String ResinArguments { get; private set; }
    public String ServiceName { get; private set; }
    public String ResinHome { get; private set; }
    public String ResinRoot { get; private set; }
    public String DisplayName { get; private set; }
    public String Command { get; private set; }
    public String ResinDataDir { get; private set; }
    public String Exe { get; private set; }
    public String Server { get; private set; }
    public String Conf { get; private set; }
    public String Log { get; private set; }
    public String JmxPort { get; private set; }
    public String DebugPort { get; private set; }
    public String WatchDogPort { get; private set; }
    public bool IsPreview { get; private set; }
    public StringCollection RawArgs { get; private set; }
    public bool DynamicServer { get; private set; }
    public bool ElasticServer { get; private set; }
    public String ElasticServerAddress { get; private set; }
    public String ElasticServerPort { get; private set; }
    public String Cluster { get; private set; }

    public ResinArgs(String cmd)
    {
      int resinIdx = cmd.IndexOf("resin.exe");
      if (resinIdx == -1)
        resinIdx = cmd.IndexOf("httpd.exe");

      if (resinIdx > 0 && cmd[0] == '"')
        Exe = cmd.Substring(1, resinIdx - 1 + 9);
      else if (resinIdx > 0)
        Exe = cmd.Substring(0, resinIdx + 9);

      StringCollection arguments = new StringCollection();
      if (Exe != null)
        arguments.Add(Exe);

      StringBuilder builder = new StringBuilder();

      int startIdx = 0;
      if (resinIdx > 0)
        startIdx = resinIdx + 9;
      for (; startIdx < cmd.Length; startIdx++)
      {
        if (cmd[startIdx] == ' ')
          break;
      }

      bool quoted = false;
      for (int i = startIdx; i < cmd.Length; i++)
      {
        char c = cmd[i];
        if ('"' == c && !quoted)
        {
          quoted = true;
        }
        else if ('"' == c && quoted)
        {
          if (builder.Length > 0)
            arguments.Add(builder.ToString());

          builder = new StringBuilder();
          quoted = false;
        }
        else if (' ' == c && quoted)
        {
          builder.Append(c);
        }
        else if (' ' == c)
        {
          arguments.Add(builder.ToString());
          builder = new StringBuilder();

        }
        else if (builder != null)
        {
          builder.Append(c);
        }
        else
        {
          builder.Append(c);
        }
      }

      parse(arguments);
    }

    public ResinArgs(String[] args)
    {
      StringCollection arguments = new StringCollection();
      arguments.AddRange(args);

      parse(arguments);
    }

    private void parse(StringCollection arguments)
    {
      RawArgs = arguments;
      StringBuilder resinArgs = new StringBuilder();

      int argsIdx = 1;
      while (argsIdx < arguments.Count)
      {
        if ("-install".Equals(arguments[argsIdx])
            || "--install".Equals(arguments[argsIdx])
            || "-install-as".Equals(arguments[argsIdx])
            || "--install-as".Equals(arguments[argsIdx]))
        {
          Console.WriteLine("Please use setup.exe to install the service.");

          Environment.Exit(1);
        }
        else if ("-remove".Equals(arguments[argsIdx])
            || "--remove".Equals(arguments[argsIdx])
            || "-remove-as".Equals(arguments[argsIdx])
            || "--remove-as".Equals(arguments[argsIdx]))
        {
          Console.WriteLine("Please use setup.exe to remove a service.");

          Environment.Exit(1);
        }
        else if ("-verbose".Equals(arguments[argsIdx]))
        {
          IsVerbose = true;
          argsIdx++;
          resinArgs.Append(" -verbose ");
        }
        else if (arguments[argsIdx].StartsWith("-J")
                 || arguments[argsIdx].StartsWith("-D")
                 || arguments[argsIdx].StartsWith("-X"))
        {
          resinArgs.Append(' ').Append(arguments[argsIdx++]).Append(' ');
        }
        else if ("-java_home".Equals(arguments[argsIdx])
                   || "-java-home".Equals(arguments[argsIdx]))
        {
          JavaHome = arguments[argsIdx + 1];

          argsIdx += 2;
        }
        else if ("-java_exe".Equals(arguments[argsIdx])
                 || "-java-exe".Equals(arguments[argsIdx]))
        {
          JavaExe = arguments[argsIdx + 1];

          argsIdx += 2;
        }

        else if ("-resin_home".Equals(arguments[argsIdx])
                 || "-resin-home".Equals(arguments[argsIdx])
                 || "--resin-home".Equals(arguments[argsIdx]))
        {
          ResinHome = arguments[argsIdx + 1];

          argsIdx += 2;
        }
        else if ("-server-root".Equals(arguments[argsIdx])
                 || "-server_root".Equals(arguments[argsIdx])
                 || "--root-directory".Equals(arguments[argsIdx])
                 || "-root-directory".Equals(arguments[argsIdx]))
        {
          ResinRoot = arguments[argsIdx + 1];

          argsIdx += 2;
        }
        else if ("-service".Equals(arguments[argsIdx]))
        {
          IsService = true;

          argsIdx++;
        }
        else if ("-console".Equals(arguments[argsIdx]))
        {
          Command = "console";

          argsIdx++;
        }
        else if ("-standalone".Equals(arguments[argsIdx]))
        {
          argsIdx++;
        }
        else if ("-server".Equals(arguments[argsIdx])
                 || "--server".Equals(arguments[argsIdx]))
        {
          Server = arguments[argsIdx + 1];

          argsIdx += 2;
        }
        else if ("-dynamic-server".Equals(arguments[argsIdx])
                 || "--dynamic-server".Equals(arguments[argsIdx]))
        {
          argsIdx += 2;
        }
        else if ("-join-cluster".Equals(arguments[argsIdx])
                  || "--join-cluster".Equals(arguments[argsIdx])
                  || "-cluster".Equals(arguments[argsIdx]) 
                  || "--cluster".Equals(arguments[argsIdx]))
        {
          DynamicServer = true;
          ElasticServer = true;
          Cluster = arguments[argsIdx + 1];

          argsIdx += 2;
        }
        else if ("-elastic-server".Equals(arguments[argsIdx])
                 || "--elastic-server".Equals(arguments[argsIdx])) {
          ElasticServer = true;
          argsIdx++;
        } 
        else if ("--elastic-server-address".Equals(arguments[argsIdx]) 
                 || "-elastic-server-address".Equals(arguments[argsIdx])) {
          ElasticServerAddress = arguments[argsIdx + 1];
          argsIdx += 2;
        } 
        else if ("--elastic-server-port".Equals(arguments[argsIdx])
                   || "-elastic-server-port".Equals(arguments[argsIdx])) {
          ElasticServerPort = arguments[argsIdx + 1];
          argsIdx += 2;
        }
        else if ("-conf".Equals(arguments[argsIdx])
                 || "--conf".Equals(arguments[argsIdx]))
        {
          Conf = arguments[argsIdx + 1];

          resinArgs.Append("-conf \"").Append(Conf).Append("\" ");

          argsIdx += 2;
        }
        else if ("-log-directory".Equals(arguments[argsIdx])
                 || "--log-directory".Equals(arguments[argsIdx]))
        {
          Log = arguments[argsIdx + 1];

          resinArgs.Append("-log-directory \"").Append(Log).Append("\" ");

          argsIdx += 2;
        }
        else if ("-jmx-port".Equals(arguments[argsIdx])
                 || "--jmx-port".Equals(arguments[argsIdx]))
        {
          JmxPort = arguments[argsIdx + 1];

          resinArgs.Append("-jmx-port ").Append(JmxPort).Append(' ');

          argsIdx += 2;
        }
        else if ("-debug-port".Equals(arguments[argsIdx])
                 || "--debug-port".Equals(arguments[argsIdx]))
        {
          DebugPort = arguments[argsIdx + 1];

          resinArgs.Append("-debug-port ").Append(DebugPort).Append(' ');

          argsIdx += 2;
        }
        else if ("-watchdog-port".Equals(arguments[argsIdx])
                 || "--watchdog-port".Equals(arguments[argsIdx]))
        {
          WatchDogPort = arguments[argsIdx + 1];

          resinArgs.Append("-watchdog-port ").Append(WatchDogPort).Append(' ');

          argsIdx += 2;
        }
        else if ("-name".Equals(arguments[argsIdx]) && IsService)
        {
          ServiceName = arguments[argsIdx + 1];

          argsIdx += 2;
        }
        else if ("-name".Equals(arguments[argsIdx]) && ! IsService)
        {
          resinArgs.Append(" -name \"").Append(arguments[argsIdx + 1]).Append("\" ");

          argsIdx += 2;
        }
        else if ("-preview".Equals(arguments[argsIdx])
                 || "--preview".Equals(arguments[argsIdx]))
        {
          IsPreview = true;
          argsIdx++;
        }
        else if (Command == null
                 && ("-h".Equals(arguments[argsIdx])
                     || "config-cat".Equals(arguments[argsIdx])
                     || "config-deploy".Equals(arguments[argsIdx])
                     || "config-ls".Equals(arguments[argsIdx])
                     || "config-undeploy".Equals(arguments[argsIdx])
                     || "console".Equals(arguments[argsIdx])
                     || "deploy".Equals(arguments[argsIdx])
                     || "deploy-copy".Equals(arguments[argsIdx])
                     || "deploy-list".Equals(arguments[argsIdx])
                     || "disable".Equals(arguments[argsIdx])
                     || "disable-soft".Equals(arguments[argsIdx])
                     || "enable".Equals(arguments[argsIdx])
                     || "gui".Equals(arguments[argsIdx])
                     || "heap-dump".Equals(arguments[argsIdx])
                     || "jmx-call".Equals(arguments[argsIdx])
                     || "jmx-dump".Equals(arguments[argsIdx])
                     || "jmx-list".Equals(arguments[argsIdx])
                     || "jmx-set".Equals(arguments[argsIdx])
                     || "jspc".Equals(arguments[argsIdx])
                     || "kill".Equals(arguments[argsIdx])
                     || "license-add".Equals(arguments[argsIdx])
                     || "list-restarts".Equals(arguments[argsIdx])
                     || "log-level".Equals(arguments[argsIdx])
                     || "passsword-encrypt".Equals(arguments[argsIdx])
                     || "passsword-generate".Equals(arguments[argsIdx])
                     || "pdf-report".Equals(arguments[argsIdx])
                     || "profile".Equals(arguments[argsIdx])
                     || "restart".Equals(arguments[argsIdx])
                     || "scoreboard".Equals(arguments[argsIdx])
                     || "shutdown".Equals(arguments[argsIdx])
                     || "start".Equals(arguments[argsIdx])
                     || "start-all".Equals(arguments[argsIdx])
                     || "start-with-foreground".Equals(arguments[argsIdx])
                     || "status".Equals(arguments[argsIdx])
                     || "stop".Equals(arguments[argsIdx])
                     || "thread-dump".Equals(arguments[argsIdx])
                     || "undeploy".Equals(arguments[argsIdx])
                     || "watchdog".Equals(arguments[argsIdx])
                     || "web-app-deploy".Equals(arguments[argsIdx])
                     || "web-app-restart".Equals(arguments[argsIdx])
                     || "web-app-start".Equals(arguments[argsIdx])
                     || "web-app-stop".Equals(arguments[argsIdx])
                     || "web-app-undeploy".Equals(arguments[argsIdx])
                     || "help".Equals(arguments[argsIdx])
                     || "version".Equals(arguments[argsIdx])
                     //
                     || "copy".Equals(arguments[argsIdx])
                     || "list".Equals(arguments[argsIdx])
                     || "deploy-start".Equals(arguments[argsIdx])
                     || "deploy-stop".Equals(arguments[argsIdx])
                     || "deploy-restart".Equals(arguments[argsIdx])
                     || "generate-password".Equals(arguments[argsIdx])
                     || "start-webapp".Equals(arguments[argsIdx])
                     || "stop-webapp".Equals(arguments[argsIdx])
                     || "restart-webapp".Equals(arguments[argsIdx])
                    )
                 )
        {
          Command = arguments[argsIdx];

          argsIdx++;
        }
        else
        {
          resinArgs.Append(' ').Append(arguments[argsIdx++]).Append(' ');
        }
      }

      ResinArguments = resinArgs.ToString();
    }
  }
}